### DISCUSSION TITLE
New in v9.0.0 - ✨ Tools can now be stacked to form Agents

### DISCUSSION THREAD

**Discussion by olimorris** [olimorris/codecompanion.nvim#392] [Announcements]
Full Changelog [here](https://github.com/olimorris/codecompanion.nvim/blob/ade8d29c94f6ad3c5b4ccf38517cec13126a6f0d/CHANGELOG.md).

## Demo

https://github.com/user-attachments/assets/04a2bed3-7af0-4c07-b58f-f644cef1c4bb

## Notes

Bit of a big release this. As I've commented on various posts in the last couple of weeks, [Aider](https://aider.chat) is a real source of inspiration. The demo video above draws directly from one of their videos on their site. In order to make this possible I've added a new `cmd_runner` tool which can run commands on your machine as well as allowing for tools to be grouped together, or stacked, in order to make agents.

A [new](https://github.com/olimorris/codecompanion.nvim/blob/ade8d29c94f6ad3c5b4ccf38517cec13126a6f0d/lua/codecompanion/config.lua#L286-L293) `@full_stack_dev` agent has been added which utilises the `cmd_runner`, `editor` and `files` tools and it's this that I used to create and output the [Snake](https://en.wikipedia.org/wiki/Snake_(video_game_genre)) game in the video.

The grouping of tools into an agent is particularly interesting to me as I foresee the user base creating their own tools to suit their workflow. In fact, I intend to start using CodeCompanion to complete coding benchmarks and share the outputs with you all.

The Co-Pilot adapter now supports `claude-3-5-sonnet` and the performance is fantastic, as you'd expect. That's the adapter and model I used in the demo video above.

Note that there are some breaking changes for this release. Notably the removal of the `code_runner` tool as the `cmd_runner` tool ultimately replaces it and wraps it in an approval mechanism.

If you like this release, I'm always open to a [coffee](https://buymeacoffee.com/olimorris) 😃 .

**Comment by antosha417** [comment]
Wow, this is fantastic! 🎉 Congratulations on the big release! 
Can't wait give it a try

**Reply by olimorris** [reply]
Would _really_ welcome some feedback on all of this. I suspect the prompts will need some tweaking to be really optimum.

I also think with more complex prompts I may need to implement some sort of queuing mechanism in the tools. But that's a good problem to have 😄.

**Comment by LintaoAmons** [comment]
Using it everyday!
Thanks bro

**Comment by iamreinder** [comment]
@olimorris I'm currently creating some agents for different workflows. I really like the plugin. For me having a chat works better than inline code completion. Thanks for the awesome work.

I was wondering (couldn't find it in docs and wasn't able to fully track it in the code of this plugin), how is a request towards the LLM build up exactly? There's a main system prompt (in config > opts), there's agents, there's tools. Are they added on top of each other or in stead of each other? This is usefull to know for creating an agent, whether you should repeat certain guidelines from the main system prompt or not, for instance. Thanks!

**Reply by olimorris** [reply]
> There's a main system prompt (in config > opts), there's agents, there's tools. Are they added on top of each other or in stead of each other

If you press `gd` in the chat buffer you can see the message stack. Tools are added before the current prompt you've written out. I cover tools in the [TOOLS.md](https://github.com/olimorris/codecompanion.nvim/blob/main/doc/TOOLS.md) file too.

Would love to hear your feedback and see what you've built.

**Reply by iamreinder** [reply]
I feel like I should have been able to figure that out 🙈 Thanks! That's a really helpful command.

I really like the plugin. Inline completions (like GH copilot does) aren't really for me. They break me out of my flow too much when I'm writing code. And I'm a bit mindful about what I leave to an LLM. I love to write code myself and be able to understand all aspects of it. So it works better for me to have an LLM as kind of a sparring partner. Therefore chat works better for my workflow. I've been using the plugin since early september. You're really rocking it. Great features. The pace has been incredible, but I hope you take care of yourself as well.

As for the agents, for now I simply use them to provide the LLM with certain preferences I have in the languages and frameworks I use. So that when it generates code for a specific kind of project, it takes my preferences into account. But I can imagine to do more powerful things with agents in the future. I'll figure it out as I go :-)

I hope you'll be having a wonderful time around Christmas and new year's eve, cheers!

**Reply by olimorris** [reply]
> Inline completions (like GH copilot does) aren't really for me. They break me out of my flow too much when I'm writing code

I am the exact same. I think the only time I use them are if users report an issue. I think they could quite easily become a crutch and after a while I won't be able to walk without them 😆.

> As for the agents, for now I simply use them to provide the LLM with certain preferences I have in the languages and frameworks I use

I've just started outlining a feature that will allow you to open a chat buffer with "knowledge" specific to certain scenarios. Perhaps you're working with Rails and you could open up a prompt which includes the full Rails docs as well as the main source files to some Gems you have installed. Right now, you'd need to use a few slash commands to add this in. It will be basic at first, just plain text being sent to the LLM but I'd love to start exploring contextual embeddings.

And have an amazing Christmas and New Years as well.
